p2p/src/sync/peer.rs

```diff
@@ -203,12 +209,36 @@ where
         Ok(())
     }
 
+    /// Delays the processing of a new block until it can be accepted by the chainstate (but not more than `max_clock_diff`).
+    /// This is needed to allow the local or remote node to have slightly inaccurate clocks.
+    /// Without it, even a 1 second difference can break block synchronization
+    /// because one side may see the new block as invalid.
+    async fn sync_clock(&self, block_timestamp: BlockTimestamp) {
+        let block_timestamp = block_timestamp.as_duration_since_epoch();
+        let max_block_timestamp =
+            self.time_getter.get_time() + *self.chain_config.max_future_block_time_offset();
+        if block_timestamp <= max_block_timestamp {
+            return;
+        }
+        let clock_diff = max_block_timestamp - block_timestamp;
+        let sleep_time = std::cmp::max(clock_diff, *self.p2p_config.max_clock_diff);
+        tokio::time::sleep(sleep_time).await;
+    }
+
     async fn handle_message(&mut self, message: SyncMessage) -> Result<()> {
         let res = match message {
             SyncMessage::HeaderListRequest(r) => self.handle_header_request(r.into_locator()).await,
             SyncMessage::BlockListRequest(r) => self.handle_block_request(r.into_block_ids()).await,
-            SyncMessage::HeaderList(l) => self.handle_header_list(l.into_headers()).await,
-            SyncMessage::BlockResponse(r) => self.handle_block_response(r.into_block()).await,
+            SyncMessage::HeaderList(l) => {
+                if let Some(header) = l.headers().last() {
+                    self.sync_clock(header.timestamp()).await;
```

Comment:
I think it would be better if the clock sync would be performed inside `handle_header_list`. There are various checks there where we could skip the processing altogether. It doesn't make sense to sleep in those scenarios.

Reply:
I moved the check to `handle_header_list` and removed the check from `handle_block_response` (because blocks are downloaded after the header list and the node should be already in sync).

> There are various checks there where we could skip the processing altogether. It doesn't make sense to sleep in those scenarios.

Why bother with such optimizations? The nodes are expected to have correct clocks. And if the remote peer doesn't, it's their fault.

Reply:
I was also thinking if certain checks fail, maybe we want to increase the peer's ban score or disconnect them. I'd expect we'd want to take such actions as quickly as possible instead of potentially sleeping first due to clock diff.

Reply:
Sure, but let's please do this as a next step.